--- cap_ffmpeg_impl.hpp	2019-10-27 21:06:02.144461471 +0100
+++ cap_ffmpeg_impl_ov.hpp	2019-11-09 10:51:30.675744233 +0100
@@ -40,17 +40,24 @@
 //
 //M*/
 
-#include "cap_ffmpeg_legacy_api.hpp"
-
-using namespace cv;
+/* ATTENTION:
+*
+* This file was generated from
+*
+* https://github.com/opencv/opencv/blob/master/modules/videoio/src/cap_ffmpeg_impl.hpp
+* commit 2ad0487cec35b08a1dbf49e98c27b05153aaa23a
+*
+* and applying the patches in cap_ffmpeg_impl_ov.patch
+*/
 
-#if !(defined(_WIN32) || defined(WINCE))
-# include <pthread.h>
-#endif
 #include <assert.h>
 #include <algorithm>
 #include <limits>
 
+#include <string>
+#include <octave/oct.h>
+#undef USE_AV_INTERRUPT_CALLBACK
+
 #ifndef __OPENCV_BUILD
 #define CV_FOURCC(c1, c2, c3, c4) (((c1) & 255) + (((c2) & 255) << 8) + (((c3) & 255) << 16) + (((c4) & 255) << 24))
 #endif
@@ -103,6 +110,15 @@
 #define CV_WARN(message) fprintf(stderr, "warning: %s (%s:%d)\n", message, __FILE__, __LINE__)
 #endif
 
+static int global_err;
+
+std::string get_last_err_msg ()
+{
+  char err_buf[80];
+  av_strerror (global_err, err_buf, 80);
+  return err_buf;
+}
+
 #if defined _WIN32
     #include <windows.h>
     #if defined _MSC_VER && _MSC_VER < 1900
@@ -475,13 +491,17 @@
 #endif
 }
 
+static bool capture_type_loaded = false;
 
-struct CvCapture_FFMPEG
+class CvCapture_FFMPEG: public octave_base_value
 {
+  public:
+    CvCapture_FFMPEG ();
+
     bool open( const char* filename );
     void close();
 
-    double getProperty(int) const;
+    //double getProperty(int) const;
     bool setProperty(int, double);
     bool grabFrame();
     bool retrieveFrame(int, unsigned char** data, int* step, int* width, int* height, int* cn);
@@ -497,6 +517,18 @@
     double  get_fps() const;
     int     get_bitrate() const;
 
+    AVRational get_sample_aspect_ratio () const
+      { return _opencv_ffmpeg_get_sample_aspect_ratio(ic->streams[video_stream]); }
+
+    const char* get_video_codec_name () const
+      {
+#if LIBAVFORMAT_BUILD > 4628
+        return _opencv_avcodec_get_name(video_st->codec->codec_id);
+#else
+        return _opencv_avcodec_get_name(video_st->codec.codec_id);
+#endif
+      }
+
     double  r2d(AVRational r) const;
     int64_t dts_to_frame_number(int64_t dts);
     double  dts_to_sec(int64_t dts) const;
@@ -531,10 +563,45 @@
 #if USE_AV_INTERRUPT_CALLBACK
     AVInterruptCallbackMetadata interrupt_metadata;
 #endif
+
+  bool is_constant (void) const
+  {
+    return true;
+  }
+  bool is_defined (void) const
+  {
+    return true;
+  }
+
+  DECLARE_OV_TYPEID_FUNCTIONS_AND_DATA
+
+  void print (std::ostream & os, bool pr_as_read_syntax = false)
+  {
+    os << "CvCapture_FFMPEG:" << std::endl;
+    if (filename)
+      os << "  filename           = " << filename << std::endl;
+    os << "  get_total_frames() = " << get_total_frames() << std::endl;
+    os << "  get_duration_sec() = " << get_duration_sec() << std::endl;
+    os << "  get_fps()          = " << get_fps() << std::endl;
+    os << "  get_bitrate()      = " << get_bitrate() << std::endl;
+    os << "  width              = " << frame.width << std::endl;
+    os << "  height             = " << frame.height << std::endl;
+    os << "  frame_number       = " << frame_number << std::endl;
+    os << "  video_codec_name   = " << get_video_codec_name () << std::endl;
+    AVRational s = get_sample_aspect_ratio ();
+    os << "  aspect_ration_num  = " << s.num << std::endl;
+    os << "  aspect_ration_den  = " << s.den << std::endl;
+  }
 };
 
+DEFINE_OV_TYPEID_FUNCTIONS_AND_DATA(CvCapture_FFMPEG, "CvCapture_FFMPEG", "CvCapture_FFMPEG");
+
+CvCapture_FFMPEG::CvCapture_FFMPEG ()
+  : octave_base_value () { init (); };
+
 void CvCapture_FFMPEG::init()
 {
+    av_register_all();
     ic = 0;
     video_stream = -1;
     video_st = 0;
@@ -899,24 +966,24 @@
       input_format = av_find_input_format(entry->value);
     }
 
-    int err = avformat_open_input(&ic, _filename, input_format, &dict);
+    global_err = avformat_open_input(&ic, _filename, input_format, &dict);
 #else
-    int err = av_open_input_file(&ic, _filename, NULL, 0, NULL);
+    global_err = av_open_input_file(&ic, _filename, NULL, 0, NULL);
 #endif
 
-    if (err < 0)
+    if (global_err < 0)
     {
         CV_WARN("Error opening file");
         CV_WARN(_filename);
         goto exit_func;
     }
-    err =
+    global_err =
 #if LIBAVFORMAT_BUILD >= CALC_FFMPEG_VERSION(53, 6, 0)
     avformat_find_stream_info(ic, NULL);
 #else
     av_find_stream_info(ic);
 #endif
-    if (err < 0)
+    if (global_err < 0)
     {
         CV_WARN("Could not find codec parameters");
         goto exit_func;
@@ -1163,7 +1230,7 @@
     return true;
 }
 
-
+#if 0
 double CvCapture_FFMPEG::getProperty( int property_id ) const
 {
     if( !video_st ) return 0;
@@ -1220,9 +1287,9 @@
     default:
         break;
     }
-
     return 0;
 }
+#endif
 
 double CvCapture_FFMPEG::r2d(AVRational r) const
 {
@@ -1361,7 +1428,7 @@
 bool CvCapture_FFMPEG::setProperty( int property_id, double value )
 {
     if( !video_st ) return false;
-
+#if 0
     switch( property_id )
     {
     case CAP_PROP_POS_MSEC:
@@ -1389,14 +1456,19 @@
     default:
         return false;
     }
-
+#endif
     return true;
 }
 
 
 ///////////////// FFMPEG CvVideoWriter implementation //////////////////////////
-struct CvVideoWriter_FFMPEG
+static bool writer_type_loaded = false;
+
+class CvVideoWriter_FFMPEG: public octave_base_value
 {
+  public:
+    CvVideoWriter_FFMPEG ();
+
     bool open( const char* filename, int fourcc,
                double fps, int width, int height, bool isColor );
     void close();
@@ -1420,8 +1492,44 @@
     int               frame_idx;
     bool              ok;
     struct SwsContext *img_convert_ctx;
+
+    const char* get_video_codec_name () const
+      {
+#if LIBAVFORMAT_BUILD > 4628
+        return _opencv_avcodec_get_name(video_st->codec->codec_id);
+#else
+        return _opencv_avcodec_get_name(video_st->codec.codec_id);
+#endif
+      }
+
+    bool is_constant (void) const
+    {
+      return true;
+    }
+    bool is_defined (void) const
+    {
+      return true;
+    }
+
+    DECLARE_OV_TYPEID_FUNCTIONS_AND_DATA
+
+    void print (std::ostream & os, bool pr_as_read_syntax = false)
+    {
+      os << "CvVideoWriter_FFMPEG:" << std::endl;
+      os << "  ok                      = " << ok << std::endl;
+      os << "  frame_width             = " << frame_width << std::endl;
+      os << "  frame_height            = " << frame_height << std::endl;
+      os << "  frame_idx               = " << frame_idx << std::endl;
+
+    // FIXME: add more properies
+    }
 };
 
+DEFINE_OV_TYPEID_FUNCTIONS_AND_DATA(CvVideoWriter_FFMPEG, "CvVideoWriter_FFMPEG", "CvVideoWriter_FFMPEG");
+
+CvVideoWriter_FFMPEG::CvVideoWriter_FFMPEG ()
+  : octave_base_value () { init (); };
+
 static const char * icvFFMPEGErrStr(int err)
 {
 #if LIBAVFORMAT_BUILD >= CALC_FFMPEG_VERSION(53, 2, 0)
@@ -1482,6 +1590,7 @@
 
 void CvVideoWriter_FFMPEG::init()
 {
+    av_register_all();
     fmt = 0;
     oc = 0;
     outbuf = 0;
@@ -2016,7 +2125,7 @@
 {
     InternalFFMpegRegister::init();
     CV_CODEC_ID codec_id = CV_CODEC(CODEC_ID_NONE);
-    int err, codec_pix_fmt;
+    int codec_pix_fmt;
     double bitrate_scale = 1;
 
     close();
@@ -2295,14 +2404,14 @@
     c->bit_rate = (int)lbit_rate;
 
     /* open the codec */
-    if ((err=
+    if ((global_err=
 #if LIBAVCODEC_VERSION_INT >= ((53<<16)+(8<<8)+0)
          avcodec_open2(c, codec, NULL)
 #else
          avcodec_open(c, codec)
 #endif
          ) < 0) {
-        fprintf(stderr, "Could not open codec '%s': %s\n", codec->name, icvFFMPEGErrStr(err));
+        fprintf(stderr, "Could not open codec '%s': %s\n", codec->name, icvFFMPEGErrStr(global_err));
         return false;
     }
 
@@ -2353,12 +2462,12 @@
 
 #if LIBAVFORMAT_BUILD >= CALC_FFMPEG_VERSION(52, 111, 0)
     /* write the stream header, if any */
-    err=avformat_write_header(oc, NULL);
+    global_err=avformat_write_header(oc, NULL);
 #else
-    err=av_write_header( oc );
+    global_err=av_write_header( oc );
 #endif
 
-    if(err < 0)
+    if(global_err < 0)
     {
         close();
         remove(filename);
@@ -2404,10 +2513,10 @@
     return capture->setProperty(prop_id, value);
 }
 
-double cvGetCaptureProperty_FFMPEG(CvCapture_FFMPEG* capture, int prop_id)
-{
-    return capture->getProperty(prop_id);
-}
+//~ double cvGetCaptureProperty_FFMPEG(CvCapture_FFMPEG* capture, int prop_id)
+//~ {
+    //~ return capture->getProperty(prop_id);
+//~ }
 
 int cvGrabFrame_FFMPEG(CvCapture_FFMPEG* capture)
 {
